src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/api/dotnet/DotnetBackend.scala

```diff
@@ -64,6 +68,23 @@ class DotnetBackend extends Logging {
     channelFuture.channel().localAddress().asInstanceOf[InetSocketAddress].getPort
   }
 
+  private[dotnet] def setCallbackClient(address: String, port: Int): Unit = {
+    callbackClient = callbackClient match {
+      case Some(_) => throw new Exception("Callback client already set.")
+      case None =>
+        logInfo(s"Connecting to a callback server at $address:$port")
+        Some(new CallbackClient(new SerDe(objectsTracker), address, port))
+    }
+  }
+
+  private[dotnet] def shutdownCallbackClient(): Unit = {
```

Comment:
Should we make `setCallbackClient` and `shutdownCallbackClient` `synchronized` since there is no guaranteed that this will be called in a single thread from the Scala point of view?

Reply:
Agree. `synchronized` was initially applied for 3.0, but for some reason I've forgot to add it for 2.3 and 2.4. I've applied as suggested for all three versions.